core/src/main/java/io/camunda/migrator/HistoryMigrator.java

```diff
@@ -320,96 +359,102 @@ private void migrateUserTasks() {
-          ProcessInstanceEntity processInstance = findProcessInstanceByLegacyId(legacyFlowNode.getProcessInstanceId());
-          if (processInstance != null) {
-            Long processInstanceKey = processInstance.processInstanceKey();
-            Long processDefinitionKey = findProcessDefinitionKey(legacyFlowNode.getProcessDefinitionId());
-            FlowNodeInstanceDbModel dbModel = flowNodeConverter.apply(legacyFlowNode, processDefinitionKey, processInstanceKey);
-            flowNodeMapper.insert(dbModel);
-            insertKeyIdMapping(legacyFlowNodeId, dbModel.flowNodeInstanceKey(), IdKeyMapper.TYPE.HISTORY_FLOW_NODE);
-            LOGGER.debug("Migration of legacy flow node with id [{}] completed.", legacyFlowNodeId);
-          } else {
-            LOGGER.debug("Migration of legacy flow node with id [{}] skipped. Process instance not yet available.",
-                legacyFlowNodeId);
+          if (!dbClient.checkExists(legacyFlowNodeId)) {
+            LOGGER.debug("Migrating legacy flow node with id: [{}]", legacyFlowNodeId);
+            ProcessInstanceEntity processInstance = findProcessInstanceByLegacyId(
+                legacyFlowNode.getProcessInstanceId());
+            if (processInstance != null) {
+              Long processInstanceKey = processInstance.processInstanceKey();
+              Long processDefinitionKey = findProcessDefinitionKey(legacyFlowNode.getProcessDefinitionId());
+              FlowNodeInstanceDbModel dbModel = flowNodeConverter.apply(legacyFlowNode, processDefinitionKey,
+                  processInstanceKey);
+              flowNodeMapper.insert(dbModel);
+              dbClient.insert(legacyFlowNodeId, legacyFlowNode.getStartTime(), dbModel.flowNodeInstanceKey(),
+                  HISTORY_FLOW_NODE);
+              LOGGER.debug("Migration of legacy flow node with id [{}] completed.", legacyFlowNodeId);
+            } else {
+              LOGGER.debug("Migration of legacy flow node with id [{}] skipped. Process instance not yet available.",
+                  legacyFlowNodeId);
+            }
           }
         });
   }
 
-  protected void insertKeyIdMapping(String id, Long key, IdKeyMapper.TYPE type) {
-    var idKeyDbModel = new IdKeyDbModel();
-    idKeyDbModel.setId(id);
-    idKeyDbModel.setInstanceKey(key);
-    idKeyDbModel.setType(type);
-    idKeyMapper.insert(idKeyDbModel);
-  }
-
   protected ProcessInstanceEntity findProcessInstanceByLegacyId(String processInstanceId) {
     if (processInstanceId == null)
       return null;
 
-    Long key = idKeyMapper.findKeyById(processInstanceId);
+    Long key = dbClient.findKeyById(processInstanceId);
     if (key == null) {
       return null;
     }
 
     return processInstanceMapper.findOne(key);
   }
 
+  protected Date getLatestStartDateByType(IdKeyMapper.TYPE type) {
+    Date latestStartDate = callApi(() -> dbClient.findLatestStartDateByType(type));
```

Comment:
This is now double wrapped into `#callApi`.

Reply:
Oops 👀